Bottom: 442a4730e18c8218c569fcd34985148ee291ca27
Top:    2bff98003ee8b810ac6d81b30d4f8cb6b34f4ad4
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-17 07:44:14 +0800

tuntap: enable bh early during processing XDP

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 7b5748f..39b23fc 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1707,22 +1707,18 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 			goto err_xdp;
 		}
 	}
+	rcu_read_unlock();
+	local_bh_enable();
 
 	skb = build_skb(buf, buflen);
-	if (!skb) {
-		rcu_read_unlock();
-		local_bh_enable();
+	if (!skb)
 		return ERR_PTR(-ENOMEM);
-	}
 
 	skb_reserve(skb, pad - delta);
 	skb_put(skb, len);
 	get_page(alloc_frag->page);
 	alloc_frag->offset += buflen;
 
-	rcu_read_unlock();
-	local_bh_enable();
-
 	return skb;
 
 err_redirect:
